api/filrewardsd/service/service.go

```diff
@@ -0,0 +1,379 @@
+	}
+	if err := cursor.Err(); err != nil {
+		return nil, fmt.Errorf("iterating cursor while building cache: %v", err)
+	}
+
+	s := &Service{
+		col:               col,
+		rewardsCache:      cache,
+		baseAttoFILReward: config.BaseAttoFILReward,
+	}
+
+	if config.AnalyticsAddr != "" {
+		s.ac, err = analytics.New(config.AnalyticsAddr)
+		if err != nil {
+			return nil, fmt.Errorf("creating analytics client: %s", err)
+		}
+	}
+
+	s.server = grpc.NewServer()
+	go func() {
+		pb.RegisterFilRewardsServiceServer(s.server, s)
+		if err := s.server.Serve(config.Listener); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
+			log.Errorf("serve error: %v", err)
+		}
+	}()
+
+	return s, nil
+}
+
+func (s *Service) ProcessAnalyticsEvent(ctx context.Context, req *pb.ProcessAnalyticsEventRequest) (*pb.ProcessAnalyticsEventResponse, error) {
+	reward := pb.Reward_REWARD_UNSPECIFIED
+	switch req.AnalyticsEvent {
+	case analyticspb.Event_EVENT_KEY_ACCOUNT_CREATED:
+		reward = pb.Reward_REWARD_FIRST_KEY_ACCOUNT_CREATED
+	case analyticspb.Event_EVENT_KEY_USER_CREATED:
+		reward = pb.Reward_REWARD_FIRST_KEY_USER_CREATED
+	case analyticspb.Event_EVENT_ORG_CREATED:
+		reward = pb.Reward_REWARD_FIRST_ORG_CREATED
+	case analyticspb.Event_EVENT_BILLING_SETUP:
+		reward = pb.Reward_REWARD_INITIAL_BILLING_SETUP
+	case analyticspb.Event_EVENT_BUCKET_CREATED:
+		reward = pb.Reward_REWARD_FIRST_BUCKET_CREATED
+	case analyticspb.Event_EVENT_BUCKET_ARCHIVE_CREATED:
+		reward = pb.Reward_REWARD_FIRST_BUCKET_ARCHIVE_CREATED
+	case analyticspb.Event_EVENT_MAILBOX_CREATED:
+		reward = pb.Reward_REWARD_FIRST_MAILBOX_CREATED
+	case analyticspb.Event_EVENT_THREAD_DB_CREATED:
+		reward = pb.Reward_REWARD_FIRST_THREAD_DB_CREATED
+	}
+	if reward == pb.Reward_REWARD_UNSPECIFIED {
+		// It is normal to get an analytics event we aren't interested in, so just return an empty result and no error.
+		return &pb.ProcessAnalyticsEventResponse{}, nil
+	}
+
+	ensureKeyRewardCache(s.rewardsCache, req.Key)
+
+	if _, exists := s.rewardsCache[req.Key][reward]; exists {
+		// This reward is already granted so bail.
+		return &pb.ProcessAnalyticsEventResponse{}, nil
+	}
+
+	rec := &rewardRecord{
+		Key:               req.Key,
+		AccountType:       req.AccountType,
+		Reward:            reward,
+		Factor:            rewardMeta[reward].factor,
+		BaseAttoFILReward: s.baseAttoFILReward,
+		CreatedAt:         time.Now(),
+	}
+
+	if _, err := s.col.InsertOne(ctx, rec); err != nil {
+		return nil, status.Errorf(codes.Internal, "inserting reward record: %v", err)
+	}
+
+	s.rewardsCache[req.Key][reward] = struct{}{}
+
+	if err := s.ac.Track(
+		ctx,
+		rec.Key,
+		rec.AccountType,
+		analyticspb.Event_EVENT_FIL_REWARD_RECORDED,
+		analytics.WithProperties(map[string]interface{}{
+			"reward":               rec.Reward,
+			"factor":               rewardMeta[rec.Reward].factor,
+			"base_atto_fil_reward": s.baseAttoFILReward,
+		}),
+	); err != nil {
+		log.Errorf("calling analytics track: %v", err)
+	}
```

Comment:
@andrewxhill here is the new event that goes to Segment whenever a new reward record is recorded.

Reply:
Decided to go with a single event, adding the specifics as properties.